In [ ]:
# Load necessary libraries
import sys
import os

sys.path.append(os.path.abspath(".."))

import pandas as pd 
from sklearn.model_selection import train_test_split

from src.preprocessing import build_preprocessor
from src.models import build_logistic_model
from src.evaluation import evaluate_model

In [ ]:
# Load the data
df = pd.read_csv("../data/WA_Fn-UseC_-Telco-Customer-Churn.csv")

# Preprocess the data
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df = df.dropna()

df['Churn'] = df['Churn'].map({'Yes': 1, 'No': 0})
df = df.drop(columns=['customerID'])

Split training and test data

In [ ]:
# Split the data
X = df.drop(columns='Churn')
y = df['Churn']

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
# Build the model
preprocessor = build_preprocessor(X)
model = build_logistic_model(preprocessor)
model.fit(X_train, y_train)

In [ ]:
# Evaluate the model
y_pred = model.predict(X_test)
y_proba = model.predict_proba(X_test)[:, 1]

evaluate_model(y_test, y_pred, y_proba)

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.89      0.87      1033
           1       0.65      0.57      0.61       374

    accuracy                           0.80      1407
   macro avg       0.75      0.73      0.74      1407
weighted avg       0.80      0.80      0.80      1407

ROC AUC Score: 0.8359290473207676
Confusion Matrix:
 [[917 116]
 [160 214]]


The model achieves strong discrimination (ROC-AUC = 0.84), but recall for churners is limited (57%), indicating that many at-risk customers are not being identified. Adjusting the classification threshold can significantly improve recall, aligning the model with business objectives focused on customer retention.